### PR TITLE
Add rule to filter unwanted CPS types

### LINKED ISSUE #1752: WS Preprocessor - Filter out unapproved asset types
**Is your feature request related to a problem? Please describe.**
To avoid launch complication, we are going to filter out unapproved asset types

**Describe the solution you'd like**
Create a preprocessor rule in simorgh that loops over all sections and all items, removing any items that aren't `STY` or `MAP`

Use a whitelist of `STY` or `MAP`, using the `cpsType` key in the items

An example of a preprocessor rule can be found [here](https://github.com/bbc/simorgh/blob/9fa9eb596a3eae1d6d6d1e87eab344e9ee9db472/src/app/lib/utilities/preprocessor/rules/topstories.js). TL;DR, take in json, split out json 

**Testing notes**
[Tester to complete]

Dev insight: Unit tests for daayyysss

### PR DESCRIPTION
**Pull request by andrew-nowak**
Resolves #1752

**Overall change:** Add a preprocessor rule that filters items with unknown and missing `cpsType`s from front page data feeds.
The current whitelist is `['STY', 'MAP']`

**Code changes:**

- Implement the rule
- The rule is *not* in the list of [default rules](https://github.com/bbc/simorgh/blob/578a2d48a0526ad02ce89d3523cb6d5ff949564c/src/app/lib/utilities/preprocessor/index.js#L4), so is not expected to run without a further code change.

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [x] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

### CODE REVIEW

File: src/app/lib/utilities/preprocessor/rules/cpstypes.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+const filterUnknownCpsTypes = data => {
+  const groups = deepGet(['content', 'groups'], data);
+
+  if (!groups) {
+    return data;
+  }
+
+  const newGroups = groups.map(g => {
+    const ng = g;
```
Comment (dr3) [inline src/app/lib/utilities/preprocessor/rules/cpstypes.js]: Do we need this line?

File: src/app/lib/utilities/preprocessor/rules/cpstypes.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+const whitelist = ['STY', 'MAP'];
+
+const filterUnknownCpsTypes = data => {
+  const groups = deepGet(['content', 'groups'], data);
+
+  if (!groups) {
+    return data;
+  }
+
+  const newGroups = groups.map(g => {
```
Comment (dr3) [inline src/app/lib/utilities/preprocessor/rules/cpstypes.js]: Also, as clean as it is, can we use words for variables, rather than letters please. Makes it a little bit easier to read quickly

**Review by dr3 — CHANGES_REQUESTED** [review]
Looks great, just a couple tweaks. Nice tests, very easy to see each edge case

File: src/app/lib/utilities/preprocessor/rules/cpstypes.js (hunk 1 above)
Comment (andrew-nowak) [reply src/app/lib/utilities/preprocessor/rules/cpstypes.js]: Not doing this causes complaints from <https://eslint.org/docs/rules/no-param-reassign>, as it was done in the [topstories preprocessor rule](https://github.com/bbc/simorgh/blob/85c8989f2370144b35e8a26ec73807e1d065d658/src/app/lib/utilities/preprocessor/rules/topstories.js#L11) (for as far as I can tell the same reason) I thought to keep the same pattern

Comment (dr3) [reply src/app/lib/utilities/preprocessor/rules/cpstypes.js]: Ahhh thats what its for :P Gotcha, can you make the names words please :P

File: src/app/lib/utilities/preprocessor/rules/cpstypes.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  if (!groups) {
+    return data;
+  }
+
+  const newGroups = groups.map(group => {
+    // avoid eslint complaints about 'no-param-reassign'
+    const newGroup = group;
+    const newItems =
+      group.items &&
```
Comment (dr3) [inline src/app/lib/utilities/preprocessor/rules/cpstypes.js]:
```suggestion
      Array.isArray(group.items) &&
```
Can you make a quick tests to check this logic please?

**Review by dr3 — CHANGES_REQUESTED** [review]
Just 1 more thing then LGTM :P

File: src/app/lib/utilities/preprocessor/rules/cpstypes.js (hunk 3 above)
Comment (andrew-nowak) [reply src/app/lib/utilities/preprocessor/rules/cpstypes.js]: Ah! good point! Didn't fix it quite as you suggested though; this would replace eg. `items: undefined` (or missing items field) with `items: false`. I went with

```js
if (Array.isArray(group.items)) {
  newGroup.items = group.items.filter(
    item => item.cpsType && whitelist.includes(item.cpsType),
  );
}
```

which should leave non-array items as they are (or not set them), allowing the group to be caught by #1770.

**Review by dr3 — APPROVED** [review]

**Review by ghost — APPROVED** [review]
Nice Tests. Great work

**Comment by david-boydell** [comment]
Nothing to manually verify here, all good for RFM.
